File: support/java/cmf/MessageBuilder.java
/* Copyright (c) 2016 Tom Zander <tomz@freedommail.ch>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package cmf;


/**
 * Message builder creates a generic message in the form of
 *    name: value
 * pairs, in a flat list.
 *
 * The unique part is that the value is typed and for variable-length structures
 * (like a string) a length is included.
 * The effect is that you can fully parse a structure without having any prior knowledge
 * of the fields, the expected content in the fields and the size.
 * You can compare this to an XML stream where some items are stored with tags or attributes
 * are unknown to the reader, without causing any effect on being able to parse them or to write
 * them out again unchanged.
 */
public class MessageBuilder
{
    public MessageBuilder(byte[] data, int position) {
        buffer = data;
        this.position = position;
    }

    public void add(int tag, long value) {
        ValueType vt;
        if (value >= 0){
            vt = ValueType.POSITIVE_NUMBER;
        } else {
            vt = ValueType.NEGATIVE_NUMBER;
            value *= -1;
        }
        write(tag, vt);
        VarInt vi = new VarInt(buffer, position);
        position += vi.serialize(value);
    }

    public void add(int tag, int value) {
        add(tag, (long) value);
    }

    // Only throws if your JRE doens't support UTF-8, which it really should
    public void add(int tag, String value) throws java.io.UnsupportedEncodingException {
        write(tag, ValueType.STRING);
        byte[] serializedData = value.getBytes("UTF-8");
        VarInt vi = new VarInt(buffer, position);
        position += vi.serialize(serializedData.length);
        System.arraycopy(serializedData, 0, buffer, position, serializedData.length);
        position += serializedData.length;
    }

    public void add(int tag, byte[] value) {
        write(tag, ValueType.BYTEARRAY);
        VarInt vi = new VarInt(buffer, position);
        position += vi.serialize(value.length);
        System.arraycopy(value, 0, buffer, position, value.length);
        position += value.length;
    }

    public void add(int tag, boolean value) {
        write(tag, value ? ValueType.BOOL_TRUE : ValueType.BOOL_FALSE);
    }

    public int getPosition() {
        return position;
    }

    private byte[] buffer = null;
    private int position = -1;

    private void write(int tag, ValueType type) {
        if (tag >= 31) { // use more than 1 byte
            byte b = (byte) (type.getValue() | 0xF8); // set the 'tag' to all 1s
            buffer[position] = b;
            VarInt vi = new VarInt(buffer, position + 1);
            position += vi.serialize(tag) + 1;
            return;
        }
        assert(tag < 32);
        byte b = (byte) tag;
        b = (byte) (b << 3);
        b += type.getValue();
        buffer[position++] = b;
    }
}

File: support/cppQt/MessageBuilder.cpp
/*
 * This file is part of the Bitcoin project
 * Copyright (C) 2014-2016 Tom Zander <tomz@freedommail.ch>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
#include "MessageBuilder.h"
#include "CMF.h"

#include <QBuffer>
#include <QDebug>

namespace {
    int write(char *data, quint32 tag, CMF::ValueType type) {
        Q_ASSERT(type < 8);
        if (tag >= 31) { // use more than 1 byte
            quint8 byte = type | 0xF8; // set the 'tag' to all 1s
            data[0] = byte;
            return CMF::serialize(data +1, tag) + 1;
        }
        Q_ASSERT(tag < 32);
        quint8 byte = tag;
        byte = byte << 3;
        byte += type;
        data[0] = byte;
        return 1;
    }
}

MessageBuilder::MessageBuilder(QIODevice *device)
    : m_device(device),
      m_ownsBuffer(false)
{
    Q_ASSERT(device);
}

MessageBuilder::MessageBuilder(QByteArray *byteArray)
    : m_device(new QBuffer(byteArray)),
      m_ownsBuffer(true)
{
    Q_ASSERT(byteArray);
    m_device->open(QIODevice::WriteOnly);
}

MessageBuilder::~MessageBuilder()
{
    if (m_ownsBuffer) {
        qobject_cast<QBuffer*>(m_device)->close();
        delete m_device;
    }
}

void MessageBuilder::add(quint32 tag, qint64 value)
{
    CMF::ValueType vt;
    if (value >= 0) {
        vt = CMF::PositiveNumber;
    } else {
        vt = CMF::NegativeNumber;
        value *= -1;
    }
    int tagSize = write(m_data, tag, vt);
    tagSize += CMF::serialize(m_data + tagSize, value);
    m_device->write(m_data, tagSize);
}

void MessageBuilder::add(quint32 tag, quint64 value)
{
    int tagSize = write(m_data, tag, CMF::PositiveNumber);
    tagSize += CMF::serialize(m_data + tagSize, value);
    m_device->write(m_data, tagSize);
}

void MessageBuilder::add(quint32 tag, const QString &value)
{
    int tagSize = write(m_data, tag, CMF::String);
    const QByteArray serializedData = value.toUtf8();
    tagSize += CMF::serialize(m_data + tagSize, serializedData.count());
    m_device->write(m_data, tagSize);
    m_device->write(serializedData);
}

void MessageBuilder::add(quint32 tag, const QByteArray &data)
{
    int tagSize = write(m_data, tag, CMF::ByteArray);
    tagSize += CMF::serialize(m_data + tagSize, data.count());
    m_device->write(m_data, tagSize);
    m_device->write(data);
}

void MessageBuilder::add(quint32 tag, bool value)
{
    int tagSize = write(m_data, tag, value ? CMF::BoolTrue : CMF::BoolFalse);
    m_device->write(m_data, tagSize);
}

void MessageBuilder::close()
{
    if (m_ownsBuffer)
        qobject_cast<QBuffer*>(m_device)->close();
}
